dnf/module/module_base.py

```diff
@@ -140,31 +140,139 @@ def install(self, module_specs, strict=True):
         if fail_safe_repo_used:
             raise dnf.exceptions.Error(_(
                 "Installing module from Fail-Safe repository is not allowed"))
-        #  Remove source packages they cannot be installed or upgraded
-        base_no_source_query = self.base.sack.query().filterm(arch__neq=['src', 'nosrc']).apply()
-        install_base_query = base_no_source_query.filter(nevra_strict=install_set_artefacts)
+        __, profiles_errors = self._install_profiles_internal(
+            install_set_artefacts, install_dict, strict)
+        if profiles_errors:
+            error_specs.extend(profiles_errors)
 
-        # add hot-fix packages
-        hot_fix_repos = [i.id for i in self.base.repos.iter_enabled() if i.module_hotfixes]
-        hotfix_packages = base_no_source_query.filter(
-            reponame=hot_fix_repos, name=install_dict.keys())
-        install_base_query = install_base_query.union(hotfix_packages)
+        if no_match_specs or error_specs or solver_errors:
+            raise dnf.exceptions.MarkingErrors(no_match_group_specs=no_match_specs,
+                                               error_group_specs=error_specs,
+                                               module_depsolv_errors=solver_errors)
 
-        for pkg_name, set_specs in install_dict.items():
-            query = install_base_query.filter(name=pkg_name)
-            if not query:
-                # package can also be non-modular or part of another stream
-                query = base_no_source_query.filter(name=pkg_name)
-                if not query:
-                    for spec in set_specs:
-                        logger.error(_("Unable to resolve argument {}").format(spec))
-                    logger.error(_("No match for package {}").format(pkg_name))
-                    error_specs.extend(set_specs)
-                    continue
-            self.base._goal.group_members.add(pkg_name)
+    def switch(self, module_specs, strict=True):
+        # :api
+        no_match_specs, error_specs, module_dicts = self.__resolve_specs_enable(module_specs)
+        # collect name of artifacts from new modules for distrosync
+        new_artefacts_names = set()
+        # collect name of artifacts from active modules for distrosync before sack update
+        active_artefacts_names = set()
+        src_arches = {"nosrc", "src"}
```

Comment:
I put it there, because I was not sure how python optimize the test `if arch in src_arches:` in for loop. In case that set is generated once, I can do it, but in case that set is generated every time, my approach is better.

Reply:
Ok. I don't know what python does here either, but if unsure I would assume it's doing the right thing and don't make the code less straightforward based on a speculation. Any performance hit also likely won't be noticeable in python anyway. But I'll leave it up to you.

Reply:
Be honest, I would like to keep it like it is.